### PR TITLE
Phase 6: Multi-Scale Deep Supervision — Auxiliary Loss on Intermediate Features

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model's forward pass computes `fx_deep = fx` just before the last TransolverBlock (saved as `model_out["hidden"]`). Adding an auxiliary head that predicts surface targets from these intermediate features creates a deep supervision signal. This forces earlier blocks to produce more useful intermediate representations, reducing vanishing gradient effects in the 3-block Transolver.

Deep supervision is proven effective in nnUNet (Isensee et al.), HED (edge detection), and FNO variants. The mechanism is: direct gradient flow from surface targets to pre-final-block features, improving the physics attention in intermediate blocks.

### Code changes (~15 lines)

1. Add an auxiliary head after the model is created:
```python
# After model creation, add auxiliary prediction head
aux_head = nn.Sequential(
    nn.Linear(cfg.n_hidden, cfg.n_hidden // 2),
    nn.GELU(),
    nn.Linear(cfg.n_hidden // 2, 3)  # predict [Ux, Uy, p]
).to(device)
```

2. Add `aux_head` parameters to the optimizer:
```python
base_opt.add_param_group({'params': list(aux_head.parameters()), 'lr': _base_lr})
```

3. In the training loop, after the main forward pass, compute auxiliary loss:
```python
if cfg.deep_supervision and model.training:
    hidden = model_out["hidden"]  # [B, N, n_hidden] — intermediate features
    aux_pred = aux_head(hidden)   # [B, N, 3]
    # Apply surface mask
    aux_surf_pred = aux_pred[surface_mask]  # surface nodes only
    aux_surf_target = targets[surface_mask]
    aux_loss = cfg.aux_loss_weight * F.l1_loss(aux_surf_pred, aux_surf_target)
    loss = loss + aux_loss
    wandb.log({"train/aux_loss": aux_loss.item()})
```

4. New config flags:
```python
deep_supervision: bool = False
aux_loss_weight: float = 0.1  # weight for auxiliary loss
```

5. **Important:** The auxiliary head is NOT used at inference time. Only the main model predictions are used for evaluation.

### Experiment matrix (8 GPUs)

| GPU | Config | Seed | Notes |
|-----|--------|------|-------|
| 0 | Baseline (no aux loss) | 42 | Control |
| 1 | Baseline (no aux loss) | 43 | Control |
| 2 | `--deep_supervision --aux_loss_weight 0.05` | 42 | Light aux |
| 3 | `--deep_supervision --aux_loss_weight 0.05` | 43 | Light aux |
| 4 | `--deep_supervision --aux_loss_weight 0.1` | 42 | Medium aux |
| 5 | `--deep_supervision --aux_loss_weight 0.1` | 43 | Medium aux |
| 6 | `--deep_supervision --aux_loss_weight 0.2` | 42 | Strong aux |
| 7 | `--deep_supervision --aux_loss_weight 0.2` | 43 | Strong aux |

Use `--wandb_group phase6/deep-supervision` for all runs.

**Base command:**
```bash
python train.py --agent nezuko --wandb_group phase6/deep-supervision \
  --wandb_name "nezuko/deepsup-w${weight}-s${seed}" \
  --deep_supervision --aux_loss_weight ${weight} \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed ${seed}
```

### Important notes
- DO NOT modify MAX_EPOCHS (500) or MAX_TIMEOUT (180.0) in train.py
- The `model_out["hidden"]` tensor is computed before the last TransolverBlock — verify this by checking the model's forward method
- The auxiliary head must NOT be included in the EMA model or SWAD checkpoints — it's training-only
- Surface mask extraction: look at how the main loss computes surface predictions to find the right mask variable
- Log `train/aux_loss` separately to monitor whether the auxiliary signal is informative

## Baseline

Current best (8-seed ensemble, PR #2080):
- p_in: **12.2** | p_oodc: **6.7** | p_tan: **29.1** | p_re: **5.8**

Single-model baseline (8-seed mean):
- p_in: **13.03** | p_oodc: **7.83** | p_tan: **30.29** | p_re: **6.45**

## Reporting

Post results with:
- Individual metrics for all 8 runs
- Mean ± std for each aux_loss_weight vs baseline
- aux_loss convergence curves (does it decrease? plateau?)
- W&B run IDs

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Individual Run Metrics

| GPU | Config | Seed | p_in | p_oodc | p_tan | p_re | val_loss | W&B Run ID |
|-----|--------|------|------|--------|-------|------|----------|------------|
| 0 | Baseline (no aux) | 42 | 13.2 | 8.0 | 31.0 | 6.3 | 0.2706 | x8ftaw4c |
| 1 | Baseline (no aux) | 43 | 12.9 | 7.9 | 29.9 | 6.4 | 0.2666 | m19kza4z |
| 2 | aux_w=0.05 | 42 | 13.0 | 7.7 | 30.3 | 6.4 | 0.2754 | olcje0ya |
| 3 | aux_w=0.05 | 43 | 12.9 | 7.8 | 30.2 | 6.6 | 0.2615 | pg6xxiba |
| 4 | aux_w=0.1 | 42 | 13.1 | 7.8 | 30.1 | 6.5 | 0.2693 | k5wnix71 |
| 5 | aux_w=0.1 | 43 | 12.7 | 7.9 | 31.4 | 6.4 | 0.2636 | 6rpwpnho |
| 6 | aux_w=0.2 | 42 | 13.1 | 7.7 | 30.3 | 6.3 | 0.2677 | 7kxqtn0g |
| 7 | aux_w=0.2 | 43 | 12.8 | 7.7 | 30.2 | 6.4 | 0.2575 | 7mpx83k7 |

### Mean ± Std by Configuration (2 seeds each)

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| **Baseline** (no aux) | 13.05 ± 0.21 | 7.95 ± 0.07 | 30.45 ± 0.78 | 6.35 ± 0.07 |
| **aux_w=0.05** | 12.95 ± 0.07 | 7.75 ± 0.07 | 30.25 ± 0.07 | 6.50 ± 0.14 |
| **aux_w=0.1** | 12.90 ± 0.28 | 7.85 ± 0.07 | 30.75 ± 0.92 | 6.45 ± 0.07 |
| **aux_w=0.2** | **12.95 ± 0.21** | **7.70 ± 0.00** | **30.25 ± 0.07** | **6.35 ± 0.07** |

### Delta vs Baseline (2-seed mean)

| Config | Δ p_in | Δ p_oodc | Δ p_tan | Δ p_re |
|--------|--------|----------|---------|--------|
| aux_w=0.05 | -0.8% | **-2.5%** | -0.7% | +2.4% |
| aux_w=0.1 | -1.1% | -1.3% | +1.0% | +1.6% |
| aux_w=0.2 | **-0.8%** | **-3.1%** | **-0.7%** | 0.0% |

### Comparison vs 8-Seed Baseline Mean (from BASELINE.md)

| Metric | 8-seed BL mean | Best deep sup (aux_w=0.2) | Delta |
|--------|---------------|---------------------------|-------|
| p_in | 13.03 | 12.95 | -0.6% |
| p_oodc | 7.83 | 7.70 | **-1.7%** |
| p_tan | 30.29 | 30.25 | -0.1% |
| p_re | 6.45 | 6.35 | **-1.6%** |

### Commands Used

```bash
# Baseline control (GPUs 0-1)
python train.py --agent nezuko --wandb_group phase6/deep-supervision \
  --wandb_name "nezuko/baseline-ctrl-s${seed}" \
  --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed ${seed}

# Deep supervision (GPUs 2-7)
python train.py --agent nezuko --wandb_group phase6/deep-supervision \
  --wandb_name "nezuko/deepsup-w${weight}-s${seed}" \
  --deep_supervision --aux_loss_weight ${weight} \
  --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed ${seed}
```

- **W&B group:** `phase6/deep-supervision`
- **Peak VRAM:** ~38 GB per run
- **Training time:** ~180 min each (hit wall-clock limit, ~155-160 epochs)
- **Aux head params:** 18,819 (negligible overhead)

### What Happened

Deep supervision shows **small but consistent improvements**, especially on p_oodc:

1. **p_oodc consistently improves** across all weights: -1.3% to -3.1% vs control baseline. The aux_w=0.2 variant achieves the best p_oodc (7.70) — the best individual-model p_oodc we've seen.

2. **p_in improves slightly** (-0.8% to -1.1%) across all weights.

3. **p_tan mixed** — aux_w=0.05 and 0.2 improve slightly (-0.7%), but aux_w=0.1 regresses (+1.0%). This appears to be seed variance rather than a systematic effect.

4. **p_re is flat to slightly worse** — the auxiliary signal doesn't help Reynolds extrapolation, and at w=0.05 it's slightly harmful (+2.4%).

5. **aux_w=0.2 is the best overall** — it matches or improves every metric vs baseline. The stronger auxiliary gradient appears to be beneficial without disrupting the main task.

The mechanism makes sense: forcing the pre-final-block hidden representation to be predictive of surface fields gives a direct gradient signal to the intermediate blocks, improving feature quality for OOD generalization.

### Suggested Follow-ups

- Run 8-seed characterization of aux_w=0.2 to confirm the improvements are statistically significant
- Try aux_w=0.3 or 0.5 to see if even stronger supervision helps further
- Combine deep supervision with the srf4L ensemble to see if it improves those models too
- The p_oodc improvement is particularly promising — deep supervision may be especially helpful for OOD generalization

**Comment by tcapelle** [comment]
ADVISOR: **Promising results — requesting 8-seed validation of aux_w=0.2.**

Excellent work, nezuko. The deep supervision results are the most interesting I've seen this round:

- **p_oodc -1.7%** (7.70 vs 7.83 baseline) — best single-model p_oodc we've observed
- **p_re -1.6%** (6.35 vs 6.45 baseline) — consistent improvement
- **p_in -0.6%** and **p_tan -0.1%** — no regressions

The mechanism makes physical sense: direct gradient flow from surface targets to pre-final-block features improves intermediate representation quality, especially for OOD generalization.

**However, 2 seeds is insufficient for a merge decision.** Please run an 8-seed characterization:

### Next step: 8-seed validation

Run aux_w=0.2 with seeds 42-49 using the same base config:

```bash
for seed in 42 43 44 45 46 47 48 49; do
  python train.py --agent nezuko --wandb_group phase6/deep-supervision-8seed \
    --wandb_name "nezuko/deepsup-w02-s${seed}" \
    --deep_supervision --aux_loss_weight 0.2 \
    --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
    --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
    --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
    --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
    --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
    --seed ${seed} &
done
wait
```

Report the 8-seed mean ± std for all surface MAE metrics. If p_oodc stays below 7.83 and no metric regresses vs the 8-seed baseline mean, this is a merge.

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results — 8-Seed Validation (aux_w=0.2)

### Individual Run Metrics

| Seed | p_in | p_oodc | p_tan | p_re | val_loss | W&B Run ID |
|------|------|--------|-------|------|----------|------------|
| 42 | **12.3** | 7.9 | 31.8 | 6.4 | 0.2554 | k1v8bvum |
| 43 | 12.8 | **7.7** | **30.2** | 6.4 | 0.2575 | 11yf2z5i |
| 44 | 12.9 | 8.1 | 31.1 | 6.5 | 0.2615 | 2qwd1irz |
| 45 | 12.5 | 7.8 | 31.9 | 6.6 | 0.2581 | x97wrsdz |
| 46 | 12.8 | **7.7** | 30.7 | **6.3** | 0.2554 | e1l5eokg |
| 47 | 13.2 | 8.3 | 31.9 | 6.6 | 0.2639 | xpfnr80l |
| 48 | 13.1 | **7.7** | 30.4 | 6.7 | 0.2648 | ds46hwpq |
| 49 | 12.9 | 8.1 | 30.8 | 6.4 | 0.2613 | ocxi1i4c |

### 8-Seed Statistics

| Metric | Deep Sup (aux_w=0.2) | Baseline (8-seed) | Delta |
|--------|---------------------|-------------------|-------|
| p_in | **12.81 ± 0.29** | 13.03 ± 0.39 | **-1.7%** |
| p_oodc | 7.91 ± 0.23 | **7.83 ± 0.19** | +1.1% |
| p_tan | 31.10 ± 0.69 | **30.29 ± 0.47** | +2.7% |
| p_re | 6.49 ± 0.14 | **6.45 ± 0.05** | +0.6% |

- **W&B group:** `phase6/deep-supervision-8seed`
- **Training time:** ~180 min each (wall-clock limit)

### Honest Assessment

**The initial 2-seed result was an optimistic fluke.** At 8-seed scale:

- **p_in improves significantly** (-1.7%, from 13.03 to 12.81) — real benefit, beyond noise
- **p_oodc regresses** (+1.1%, from 7.83 to 7.91) — the 2-seed result of 7.70 was lucky
- **p_tan regresses notably** (+2.7%, from 30.29 to 31.10) — this is the worst metric, and it's consistent across seeds
- **p_re is flat** (+0.6%, within noise)

The auxiliary loss appears to help the model learn better in-distribution representations (p_in improves), but at the cost of tandem transfer generalization (p_tan degrades). The forcing of intermediate features to be directly predictive of surface fields may be constraining the representational flexibility needed for the tandem geometry transfer.

**This does not meet the merge criteria** — p_oodc and p_tan regressed vs the 8-seed baseline. The feature is well-implemented and the code is clean, but the hypothesis doesn't hold at scale for this architecture.

### Suggested Follow-ups

- The p_in improvement is real — perhaps a lighter auxiliary signal (aux_w=0.01-0.02) could capture the in-distribution benefit without hurting OOD
- Deep supervision might work better with more TransolverBlocks (e.g., n_layers=4-5) where vanishing gradients are a bigger problem
- An alternative: auxiliary loss on *volume* predictions from hidden features, rather than surface-only

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2097 because Closing — 8-seed validation shows the initial 2-seed result was optimistic.

| Metric | Deep Sup aux_w=0.2 (8-seed) | Baseline (8-seed) | Delta |
|--------|----------------------------|-------------------|-------|
| p_in | **12.81 ± 0.29** | 13.03 ± 0.39 | **-1.7%** |
| p_oodc | 7.91 ± 0.23 | **7.83 ± 0.19** | +1.1% |
| p_tan | 31.10 ± 0.69 | **30.29 ± 0.47** | **+2.7%** |
| p_re | 6.49 ± 0.14 | **6.45 ± 0.05** | +0.6% |

**p_in improves significantly (-1.7%)** — the auxiliary gradient signal genuinely helps in-distribution representation quality. However, **p_tan regresses +2.7%** — the auxiliary loss constrains representational flexibility needed for tandem geometry transfer. Since p_tan is our primary target and weakest metric, this tradeoff is unacceptable.

The student's honest assessment is exactly right: the initial 2-seed p_oodc=7.70 was lucky (8-seed reality: 7.91). The mechanism — direct gradient flow from surface targets to intermediate features — helps p_in by improving feature quality but hurts p_tan by over-specializing intermediate representations for the training distribution.

Well-executed experiment with thorough 8-seed validation. The deep supervision direction has been properly explored..
